frontend/src/metabase/admin/performance/components/StrategyEditorForDatabases.tsx

```diff
@@ -0,0 +1,328 @@
+    ]);
+
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+      if (nextLocation) {
+        dispatch(push(nextLocation));
+      }
+    }, [dispatch, nextLocation]);
+
+    /** Update the targetId (the id of the currently edited model) but confirm if the form is unsaved */
+    const safelyUpdateTargetId: SafelyUpdateTargetId = (
+      newTargetId,
+      isFormDirty,
+    ) => {
+      if (targetId === newTargetId) {
+        return;
+      }
+      const update = () => setTargetId(newTargetId);
+      if (isFormDirty) {
+        askBeforeDiscardingChanges(update);
+      } else {
+        update();
+      }
+    };
+
+    useEffect(() => {
+      if (!canOverrideRootStrategy && targetId === null) {
+        setTargetId(rootId);
+      }
+    }, [canOverrideRootStrategy, targetId]);
+
+    const saveStrategy = useCallback(
+      async (values: Strategy) => {
+        if (targetId === null) {
+          return;
+        }
+
+        const baseConfig: Pick<Config, "model" | "model_id"> = {
+          model: targetId === rootId ? "root" : "database",
+          model_id: targetId,
+        };
+
+        const otherConfigs = configs.filter(
+          config => config.model_id !== targetId,
+        );
+        if (values.type === "inherit") {
+          await CacheConfigApi.delete(baseConfig, { hasBody: true }).then(
+            () => {
+              setConfigs(otherConfigs);
+            },
+          );
+        } else {
+          const validFields = getFieldsForStrategyType(values.type);
+          const newStrategy = pick(values, validFields) as Strategy;
+
+          const validatedStrategy =
+            Strategies[values.type].validateWith.validateSync(newStrategy);
+
+          const newConfig = {
+            ...baseConfig,
+            strategy: validatedStrategy,
+          };
+
+          await CacheConfigApi.update(newConfig).then(() => {
+            setConfigs([...otherConfigs, newConfig]);
+          });
+        }
+      },
+      [configs, targetId],
+    );
+    const formPanelRef = useRef<HTMLDivElement>(null);
+    const [formPanelHasVerticalScrollbar, setFormPanelHasVerticalScrollbar] =
+      useState(false);
+
+    useEffect(() => {
+      const formPanel = formPanelRef.current;
+      if (!formPanel) {
+        return;
+      }
+      setFormPanelHasVerticalScrollbar(
+        formPanel.scrollHeight > formPanel.clientHeight,
+      );
+    }, [
+      formPanelRef.current?.scrollHeight,
+      formPanelRef.current?.clientHeight,
+      formPanelRef,
+      setFormPanelHasVerticalScrollbar,
+    ]);
+
+    const showLoadingSpinner = useDelayedLoadingSpinner();
+    const error = databasesResult.error || configsResult.error;
+    const loading = databasesResult.isLoading || configsResult.loading;
+
+    if (error || loading) {
+      return showLoadingSpinner ? (
+        <LoadingAndErrorWrapper error={error} loading={loading} />
+      ) : (
+        <></>
+      );
+    }
```

Comment:
This should be it's own component. Even if you don't have time to make this a generic component for everyone to use, it feels quite unrelated to strategies being edited so moving it out would be nice.

Reply:
Hah, I did already move it out, but I forgot